File: app/src/main/java/com/quigglesproductions/secureimageviewer/glide/ChecksumSignature.java
package com.quigglesproductions.secureimageviewer.glide;

import androidx.annotation.NonNull;

import com.bumptech.glide.load.Key;
import com.quigglesproductions.secureimageviewer.checksum.FileChecksum;

import java.nio.ByteBuffer;
import java.security.MessageDigest;

public class ChecksumSignature implements Key {
    private final FileChecksum checksum;

    public ChecksumSignature(FileChecksum fileChecksum) {
        this.checksum = fileChecksum;
    }

    @Override
    public boolean equals(Object o) {
        if (o instanceof ChecksumSignature) {
            ChecksumSignature other = (ChecksumSignature) o;
            return checksum == other.checksum;
        }
        return false;
    }

    @Override
    public int hashCode() {
        return checksum.hashCode();
    }
    @Override
    public void updateDiskCacheKey(@NonNull MessageDigest messageDigest) {
        messageDigest.update(checksum.getChecksumString().getBytes(CHARSET));
    }
}

File: app/src/main/java/com/quigglesproductions/secureimageviewer/glide/ChecksumSignature.kt
package com.quigglesproductions.secureimageviewer.glide

import com.bumptech.glide.load.Key
import com.quigglesproductions.secureimageviewer.checksum.FileChecksum
import java.security.MessageDigest

class ChecksumSignature(private val checksum: FileChecksum) : Key {
    override fun equals(other: Any?): Boolean {
        if (other is ChecksumSignature) {
            return checksum === other.checksum
        }
        return false
    }

    override fun hashCode(): Int {
        return checksum.hashCode()
    }

    override fun updateDiskCacheKey(messageDigest: MessageDigest) {
        messageDigest.update(checksum.checksumString.toByteArray(Key.CHARSET))
    }
}
